src/libmeasurement_kit/ooni/bouncer_impl.cpp

```diff
@@ -0,0 +1,111 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "bouncer_impl.hpp"
+
+namespace mk {
+namespace ooni {
+namespace bouncer {
+
+
+ErrorOr<Var<BouncerReply>> BouncerReply::create(std::string data, Var<Logger> logger) {
+    Var<BouncerReply> reply(new BouncerReply);
+    try {
+        reply->response = json::parse(data);
+        if (reply->response.find("error") != reply->response.end()) {
+            if (reply->response["error"] == "collector-not-found") {
+                logger->warn(
+                        "collector not found for the requested test");
+                return BouncerCollectorNotFoundError();
+            }
+            if (reply->response["error"] == "invalid-request") {
+                logger->warn("invalid request sent to the bouncer");
+                return BouncerInvalidRequestError();
+            }
+            // I assume that if we receive a json with the key "error"
+            // then the json has an unknown schema and we should not
+            // continue to process it
+            logger->warn("bouncer generic error");
+            return BouncerGenericError();
+        }
+        if (reply->response["net-tests"].empty()) {
+            logger->warn("generic bouncer error");
+            return BouncerTestHelperNotFoundError();
+        }
+    } catch (std::invalid_argument &) {
+        return JsonParseError();
+    } catch (std::out_of_range &) {
+        return JsonKeyError();
+    } catch (std::domain_error &) {
+        return JsonDomainError();
+    }
```

Comment:
@AntonioLangiu, in any case, I am fine with exceptions used like this. What makes me uncomfortable is mixing exceptions usage and calling callbacks (especially because also the callbacks may except).